### PR TITLE
World Hopper plugin - Type filter

### PR DESCRIPTION
**Pull request by Blackberry0Pie**
Because why not.
example config:
config
example plugin panel
league_example

### CODE REVIEW

**Comment by dekvall** [comment]
Related to #435

**Comment by Adam-** [comment]
1) I think it would be better if this could show potentially more than one type instead of just one type. Maybe just make an option per type
2) I see that this doesn't affect the world hopper next/prev feature, which it probably should

**Comment by Blackberry0Pie** [comment]
Updated to have a filter for each type.
Added a toggle for filter exclusivity(not showing world types that are not selected).
Changed behavior of world hopper regarding {skill total, bounty, and LMS} worlds when exclusivity is true.

File: runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperConfig.java [hunk 1]
```diff
@@ -125,11 +125,110 @@ default SubscriptionFilterMode subscriptionFilter()
...
+	@ConfigItem(
+		keyName = "PVPTypeFilter",
+		name = "Show PVP",
+		description = "Consider PVP worlds",
+		position = 8
+	)
+	default boolean PVPTypeFilter()
+{
+	return false;
+}
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperConfig.java]: These lines should be indented one level further.

**Comment by Blackberry0Pie** [comment]
added config section as per request

File: runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldSwitcherPanel.java [hunk 2]
```diff
@@ -247,6 +266,67 @@ void populate(List<World> worlds)
 					break;
 			}
 
+			if (exclusiveTypeFilter)
+			{
+				if (!world.getTypes().contains(WorldType.PVP) &&
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldSwitcherPanel.java:271]: Our code convention for multiline boolean operations are to put the operators at the start of continued lines like so:

```java
a
	&& b
	&& c;
```

Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldSwitcherPanel.java:271]: Additionally, this check for `world.getTypes()` containing at least one of the many types here can be implemented more cleanly using [`Collections.disjoint(A, B)`](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#disjoint-java.util.Collection-java.util.Collection-)

File: runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperConfig.java [hunk 3]
```diff
@@ -129,18 +139,127 @@ default boolean menuOption()
...
+	)
+	default boolean leagueTypeFilter()
+	{
+		return true;
+	}
+
+	@ConfigItem(
+		keyName = "exclusiveTypeFilter",
+		name = "Exclusive filtering",
+		description = "Only show selected world types",
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperConfig.java:249]: This is not clear enough to be understandable. Without looking at the code implemented using this toggle, I cannot understand what behavior this changes.

File: runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperConfig.java [hunk 4]
```diff
@@ -30,11 +31,20 @@
...
+import net.runelite.client.config.ConfigSection;
 import net.runelite.client.config.Keybind;
 
 @ConfigGroup(WorldHopperConfig.GROUP)
 public interface WorldHopperConfig extends Config
 {
+	@ConfigSection(
+		name = "World Type Filtering",
+		description = "Filtering options for world types",
+		position = -1,
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperConfig.java:43]: Maybe this should appear at the end of the config list?

File: runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperConfig.java (hunk 3 above)
Comment (Blackberry0Pie) [reply runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperConfig.java]: Requested changes done. Hopefully this description is now understandable.

File: runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldSwitcherPanel.java [hunk 5]
```diff
@@ -230,23 +253,88 @@ void populate(List<World> worlds)
...
 					{
 						continue;
 					}
 					break;
 			}
 
+			if (exclusiveTypeFilter)
+			{
+				//this removes worlds that have no type as they will not match any filter(s)
+				EnumSet<WorldType> wt = EnumSet.allOf(WorldType.class);
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldSwitcherPanel.java:277]: No need for the shorthand, give this variable a clearer name. Maybe `allWorldTypes`?

File: runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldSwitcherPanel.java [hunk 6]
```diff
@@ -230,23 +253,88 @@ void populate(List<World> worlds)
...
+					if (!types.contains(WorldType.MEMBERS))
 					{
 						continue;
 					}
 					break;
 			}
 
+			if (exclusiveTypeFilter)
+			{
+				//this removes worlds that have no type as they will not match any filter(s)
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldSwitcherPanel.java:276]: This should be above the `Collections.disjoint()` line, not the line that defines the set of all world types.

**Comment by Nightfirecat** [comment]
This still doesn't seem to affect the prev/next world hop keybind. This was tested with exclusive filtering on, and only "Show Skill Total" selected under the filtering section.

**Comment by Nightfirecat** [comment]
Now that there's been a pretty significant other change (the last commit looks purely like a separate bugfix?) it would be helpful to squash and rebase this history into atomic commits for these changes--one for the panel bugfix, and the second for the type filter.

**Comment by Adam-** [comment]
added in db7ae8d62d57f421ba72566be8bd777a8a76452a

**Pull request closed without merging**